compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,344 @@
+          val stparams =
+            if (cls.cls.typeParams.nonEmpty)
+              Some(cls.cls.typeParams.sscope)
+            else None
+          val sparents = cls.parents.map(_.toSemanticType)
+          val sself = cls.selfType.toSemanticType
+          val decls = cls.decls.toList.sscope
+          s.ClassSignature(stparams, sparents, sself, Some(decls))
+
+        case TypeBounds(lo, hi) =>
+          // for `type X[T] = T` is equivalent to `[T] =>> T`
+          def tparams(tpe: Type): (Type, List[Symbol]) = tpe match {
+            case lambda: HKTypeLambda =>
+              val paramSyms = lambda.paramNames.flatMap { paramName =>
+                paramRefSymtab.get((lambda, paramName))
+              }
+              (lambda.resType, paramSyms)
+            case _ => (tpe, Nil)
+          }
+          val (loRes, loParams) = tparams(lo)
+          val (hiRes, hiParams) = tparams(hi)
+          val params = (loParams ++ hiParams).distinctBy(_.name)
+          val slo = loRes.toSemanticType
+          val shi = hiRes.toSemanticType
+          val stparams = params.sscope
+          s.TypeSignature(Some(stparams), slo, shi)
+
+        case other =>
+          s.ValueSignature(
+            other.toSemanticType
+          )
+      }
+      loop(tpe)
+
+    private def toSemanticType(using LinkMode, SemanticSymbolBuilder, Context): s.Type =
+      import ConstantOps._
+      def loop(tpe: Type): s.Type = tpe match {
+        case ExprType(tpe) =>
+          val stpe = loop(tpe)
+          s.ByNameType(stpe)
+
+        case TypeRef(pre, sym: Symbol) =>
+          val spre = if(tpe.hasTrivialPrefix) s.Type.Empty else loop(pre)
+          val ssym = sym.symbolName
+          s.TypeRef(spre, ssym, Seq.empty)
+
+        case TermRef(pre, sym: Symbol) =>
+          val spre = if(tpe.hasTrivialPrefix) s.Type.Empty else loop(pre)
+          val ssym = sym.symbolName
+          s.SingleType(spre, ssym)
+
+        case tref: ParamRef =>
+          val key = (tref.binder, tref.paramName)
+          paramRefSymtab.get(key) match {
+            case Some(ref) =>
+              val ssym = ref.symbolName
+              tref match {
+                case _: TypeParamRef => s.TypeRef(s.Type.Empty, ssym, Seq.empty)
+                case _: TermParamRef => s.SingleType(s.Type.Empty, ssym)
+              }
+            case None => // shouldn't happen
+              s.Type.Empty
+          }
+
+        case ThisType(TypeRef(_, sym: Symbol)) =>
+          s.ThisType(sym.symbolName)
+
+        case SuperType(thistpe, supertpe) =>
+          val spre = loop(thistpe.typeSymbol.info)
+          val ssym = supertpe.typeSymbol.symbolName
+          s.SuperType(spre, ssym)
+
+        // val clazzOf = classOf[...]
+        case ConstantType(const) if const.tag == core.Constants.ClazzTag =>
+          loop(const.typeValue)
+
+        case ConstantType(const) =>
+          s.ConstantType(const.toSemanticConst)
+
+        case rt @ RefinedType(parent, name, info) =>
```

Comment:
refinements will also be empty (in the upper bound of type param `S`, if you implement the change to `PPrint.scala`) for this example:
```scala
trait SpecialRefinement {
  def pickOne(as: String*): Option[Any]
}

class PickOneRefinement_1[S <: SpecialRefinement { def pickOne(as: String*): Option[String] }] {
  def run(s: S, as: String*): Option[String] = s.pickOne(as:_*)
}
```

Reply:
I found we fail to retrieve the symbol for `pickOne` (of `SpecialRefinement { def pickOne(...) }`) because we don't traverse and generate the semanticdb symbol occurrence for `pickOne`.

For example, the `.expect.scala` for `PickOneRefinement_1` looks like:

```scala
trait SpecialRefinement/*<-example::SpecialRefinement#*/ {
  def pickOne/*<-example::SpecialRefinement#pickOne().*/(as/*<-example::SpecialRefinement#pickOne().(as)*/: String/*->scala::Predef.String#*/*): Option/*->scala::Option#*/[Any/*->scala::Any#*/]
}

class PickOneRefinement_1/*<-example::PickOneRefinement_1#*/[S/*<-example::PickOneRefinement_1#[S]*/ <: SpecialRefinement { def pickOne(as: String*): Option[String] }] {
  def run/*<-example::PickOneRefinement_1#run().*/(s/*<-example::PickOneRefinement_1#run().(s)*/: S/*->example::PickOneRefinement_1#[S]*/, as/*<-example::PickOneRefinement_1#run().(as)*/: String/*->scala::Predef.String#*/*): Option/*->scala::Option#*/[String/*->scala::Predef.String#*/] = s/*->example::PickOneRefinement_1#run().(s)*/.pickOne/*->example::SpecialRefinement#pickOne().*/(as/*->example::PickOneRefinement_1#run().(as)*/:_*)
}
```

Because the `TypeDef` for `S` looks like, and we don't traverse the `pickOne` inside `TypeTree`.

```scala
TypeDef(
  S,
  TypeTree[TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),RefinedType(TypeRef(ThisType(TypeRef(NoPrefix,module class example)),trait SpecialRefinement),pickOne,MethodType(List(as), List(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class <repeated>),List(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class scala)),object Predef),type String)))), AppliedType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Option),List(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class scala)),object Predef),type String))))))]
)
```

Reply:
Maybe we can retrieve the symbol of `pickOne` using `SymDenotaiton#member`

Reply:
Or maybe we can traverse inside the TypeTree and construct a symbol table associate between name and TypeTree?

Reply:
`rt.member` works for this case :tada:
https://github.com/lampepfl/dotty/pull/12885/commits/332609f7adb814c813ec3addfb78e896271f3703

(However, I actually don't understand, which case it works and which case it doesn't...)